Allow dependency resolve rules and substitution rules to provide descriptive reason for selection

melix opened the pull request:
### Context

See #3719

Custom reasons are visible when running the dependency insight report:

![substitution_360](https://user-images.githubusercontent.com/316357/34059829-16e06394-e1e1-11e7-8cd0-d05d60f4d82b.gif)

bigdaz commented:
@melix @jjohannes I didn't get around to reviewing this one.

A couple of quick comments:
- it would be great if we could avoid adding more inconsistency to the DSL. (like `withDescription()` vs `because`). But my DSL-design skills are below par.
- Seems like we missed adding support for `DependencySubstitution.useTarget()`

melix commented:
FTR I used `because` only where the DSL was already not on par with what we have elsewhere, so that it reads better. So this is an _intentional_ inconsistency on a legacy DSL. I'm open to alternative names, but didn't find something that would fit the 2 contexts without being ambiguous or not well suited.

bamboo commented:
I like `because` because it reads more like prose than `withReason` or `withDescription`. The inconsistency bothers me a bit too. My preference would be to rename `withDescription` to `because`and make `useVersion` and `useTarget` part of the builder interface so one could write something like:

```kotlin
resolutionStrategy.eachDependency {
    when (requested.name) {
       "foo" -> useVersion("2.0").because("I am in control")
       "bar" -> useTarget("org.test:bar:2.0").because("why not?")
       else  -> useVersion(requested.version)
    }
}
```

melix commented:
@bamboo thanks for your input. There are 2 things to consider here:

1. we're trying to minimize the changes to an API that is going to be superceded by the component metadata rules we're implementing
2. we want to keep binary compatibility

Obviously 2. is stronger than 1, and that's the reason we can write:

`withDescription(...).useTarget(...)` but not `useTarget(...).withDescription(...)` (because changing the return type is a client breaking binary change).

However if you think that doesn't read too bad, I'm not against using:

`because(...).useTarget(...)`

bamboo commented:
> because changing the return type is a client breaking binary change

Got it.

I like `because(...).useTarget(...)` better than `withDescription(...).useTarget(...)`.

jjohannes commented on subprojects/diagnostics/src/integTest/groovy/org/gradle/api/tasks/diagnostics/DependencyReportTaskIntegrationTest.groovy:
```diff
@@ -848,6 +849,55 @@ api (n)
+"""
+    }
+
+    @NotYetImplemented
```
:o: I would prefer to have this ` @ToBeImplemented` with an adjusted assertion. To not forget about this test. E.g. if we implement this with slightly different formatting, this test will still fail without us noticing.

melix replied:
I have not been personally convinced that `@ToBeImplemented` is any better: https://gradle.slack.com/archives/C13AA53CK/p1505937273000344

melix commented:
I have rebased the changes, moved from `withDescription` to `because` consistently, and added support for `DependencySubstitution.useTarget()`. It's ready for its second round of review!
